docs/proposals/20220717-resourcedistribution-generator.md

```diff
@@ -0,0 +1,287 @@
+
+## Proposal
+
+### API Definition
+
+```go
+package main
+
+type plugin struct{
+  rf        *resmap.Factory
+  ldr       ifc.Loader
+  Name      string `json:"name,omitempty" yaml:"name,omitempty"`
+  ResourceArgs `json:",inline,omitempty" yaml:",inline,omitempty"`
+  TargetsArgs `json:",inline,omitempty" yaml:",inline,omitempty"`	
+  
+  // Options for the resourcedistribution.
+  // GeneratorOptions same as configmap and secret generator Options
+  Options *types.GeneratorOptions `json:"options,omitempty" yaml:"options,omitempty"`
+
+  // Behavior of generated resource, must be one of:
+  //   'create': create a new one
+  //   'replace': replace the existing one
+  //   'merge': merge with the existing one
+  Behavior string `json:"behavior,omitempty" yaml:"behavior,omitempty"`
+}
+
+// ResourceArgs contain arguments for the resource to be distributed.
+type ResourceArgs struct {
+  // Name of the resource to be distributed.
+  DistributionName string `json:"distributionName,omitempty" yaml:"distributionName,omitempty"`
+  
+  // KvPairSources defines places to obtain key value pairs.
+  // same as configmap and secret generator KvPairSources
+  types.KvPairSources `json:",inline,omitempty" yaml:",inline,omitempty"`
+  
+  // Options for the resource to be distributed.
+  // GeneratorOptions same as configmap and secret generator Options
+  DistributionOptions *types.GeneratorOptions `json:"distributionOptions,omitempty" yaml:"distributionOptions,omitempty"`
+
+  // Type of the secret. It can be "Opaque" (default), or "kubernetes.io/tls".
+  //
+  // If type is "kubernetes.io/tls", then "literals" or "files" must have exactly two
+  // keys: "tls.key" and "tls.crt"
+  Type string `json:"type,omitempty" yaml:"type,omitempty"`
+}
+
+// TargetsArgs defines places to obtain target namespace args.
+type TargetsArgs struct {
+  // AllNamespaces if true distribute all namespaces
+  AllNamespaces bool `json:"allNamespaces,omitempty" yaml:"allNamespaces,omitempty"`
+  
+  // ExcludedNamespaces is a list of excluded namespaces name.
+  ExcludedNamespaces []string `json:"excludedNamespaces,omitempty" yaml:"excludedNamespaces,omitempty"`
+  
+  // IncludedNamespaces is a list of included namespaces name.
+  IncludedNamespaces []string `json:"includedNamespaces,omitempty" yaml:"includedNamespaces,omitempty"`
+ 
+  // NamespaceLabelSelector for the generator.
+  NamespaceLabelSelector *metav1.LabelSelector `json:"namespaceLabelSelector,omitempty" yaml:"namespaceLabelSelector,omitempty"`
+}
+```
+
+### NameSuffixHash
+If both resourcedistribution and resource's names are suffixed with hash values, they will have the same hash value.
+
+### Weakness
+
+Using the go plugin requires that the main program and the plugin program have the same build environment. So, End users using shared plugins must compile both kustomize and plugin.
```

Comment:
Is it possible to use the [Containerized KRM Function](https://kubectl.docs.kubernetes.io/guides/extending_kustomize/containerized_krm_functions/) method? Does this mean that the user does not need to compile extra